src/components/Navbar.tsx

```diff
@@ -0,0 +1,183 @@
+import React, {
+  useCallback,
+  useContext,
+  useEffect,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from "react";
+import { IconDefinition } from "@fortawesome/free-regular-svg-icons";
+import { faCircle } from "@fortawesome/free-solid-svg-icons";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import style from "../variables.module.scss";
+import "./navbar.scss";
+import { ThemeContext, ThemeProvider } from "../theme";
+import { isNavbarItem } from "../utils";
+
+const NavbarContext = React.createContext<{
+  selectedId?: number;
+  setSelectedId?: (id?: number) => void;
+  registerItem?: (item: React.RefObject<HTMLLIElement>) => void;
+}>({});
+
+type NavbarItemProps = {
+  icon: IconDefinition;
+  title: string;
+  onSelect: ({ title, id }: { title: string; id?: number }) => void;
+  id?: number; // TODO: find out how not to expose this prop
+};
+
+export function NavbarItem({ icon, title, onSelect, id }: NavbarItemProps) {
+  const theme = useContext(ThemeContext);
+  const { selectedId, setSelectedId, registerItem } = useContext(NavbarContext);
+  const itemRef = useRef<HTMLLIElement>(null);
+
+  const handleClick = () => {
+    // If is already selected, then deselect
+    if (id === selectedId) {
+      setSelectedId?.(undefined);
+    } else {
+      setSelectedId?.(id);
+      onSelect({ title, id });
+    }
+  };
+
+  const handleKeyDown = (event: React.KeyboardEvent) => {
+    switch (event.key) {
+      case "Enter":
+      case " ":
+        // If is already selected, then deselect
+        if (id === selectedId) {
+          setSelectedId?.(undefined);
+        } else {
+          setSelectedId?.(id);
+          onSelect({ title, id });
+        }
+        break;
+      default:
+        break;
+    }
+  };
+
+  useEffect(() => {
+    registerItem?.(itemRef);
+  }, [registerItem]);
+
+  const isSelected = id === selectedId;
+
+  return (
+    <li
+      tabIndex={0}
+      className="navbar-item"
+      onClick={handleClick}
+      onKeyDown={handleKeyDown}
+      ref={itemRef}
+    >
+      <FontAwesomeIcon
+        className="navbar-item-icon"
+        icon={icon}
+        size="2x" // this should be a param
+        color={
+          isSelected
+            ? style.primary
+            : theme?.colorMode === "dark"
+            ? style.lightBackground
+            : style.darkBackground
+        }
+      />
+      <div className={`navbar-item-title ${isSelected ? "selected" : ""}`}>
+        <span className="navbar-item-title-text">{title}</span>
+      </div>
+    </li>
+  );
+}
+
+type NavbarProps = {
+  label: string;
+  children: React.ReactNode;
+};
+
+export function Navbar(props: NavbarProps) {
+  const [selectedId, setSelectedId] = useState<number>();
+  const [items, setItems] = useState<React.RefObject<HTMLLIElement>[]>([]);
+  const selectedIndicatorRef = useRef<HTMLDivElement>(null);
+
+  // TODO: refactor this monstrosity (maybe ask Sandrina to give some tips)
```

Comment:
_make it work, then make it better_.  Indeed it can be reduced, quite a lot. First, I'd recommend you to:
- extract this logic to its own hook eg `useDotAnimation`
- Avoid accessing the DOM directly. Instead store the styles in useState and pass them down as: `<div ref={selectedIndicatorRef} style={dotStyles}> ...</div>` 

I'll review this part again later tonight/tomorrow, as I have to jump in a call now